apps/aecore/lib/aecore/structures/data_tx.ex

```diff
@@ -32,49 +43,65 @@ defmodule  Aecore.Structures.DataTx do
   defstruct [:type, :payload, :from_acc, :fee, :nonce]
   use ExConstructor
 
-  @spec init(atom(), map(), binary(), integer(), integer()) :: {:ok, DataTx.t()}
+  @spec init(txType(), payload(), binary(), integer(), integer()) :: DataTx.t()
   def init(type, payload, from_acc, fee, nonce) do
-    %__MODULE__{type: type,
+    %DataTx{type: type,
                 payload: type.init(payload),
                 from_acc: from_acc,
                 fee: fee,
                 nonce: nonce}
   end
 
-  @spec is_valid(DataTx.t()) :: :ok | {:error, reason()}
-  def is_valid(%__MODULE__{type: type, payload: payload, fee: fee}) do
+  @spec is_valid?(DataTx.t()) :: boolean()
+  def is_valid?(%DataTx{type: type, payload: payload, fee: fee}) do
```

Comment:
why do we need to check if it is bigger than 0 before passing to the type specific check? shouldn't the `type.is_valid?()` eliminate all 0 fees anyways?

Reply:
Because then we can end the function earlier if the fee is 0 and don't go deeper into the code

Reply:
what if there is a transaction that doesn't include a fee as rule? e.g. the coinbase transaction?

Reply:
The coinbase transaction sets fee: 0, so it will be catched with fee >= 0

Reply:
oh yes, it's greater-equals, but I think its not necessary, just increases complexity

Reply:
I'd agree to keep it if we turn the check around, make it fail if it is smaller 0 in semantics

Reply:
Not necessarily, in my opinion it gives separation of the validation checks, and it makes the checks in steps. That way we reduce unecessary computing (eg check if the specific transaction is valid before checking if the data transaction fields are valid beforehand). The fee is a type inside the DataTx, therefore it should be checked in DataTx module, the rest checks are inside the `type` in this case `:value`

Reply:
yes I agree, but make the case you check correct in a semantic mannor, so that by reading the code  this allows ANY transaction to be fee > 0